cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/HttpIdleStateHandler.java

```diff
@@ -0,0 +1,302 @@
+ * or {@link Timer#stop()} when your application shuts down.
+ *
+ */
+@ChannelHandler.Sharable
+public class HttpIdleStateHandler extends SimpleChannelHandler
+  implements LifeCycleAwareChannelHandler, ExternalResourceReleasable {
+
+  private final Timer timer;
+  private final long requestIdleTimeMillis;
+
+  /**
+   * Creates a new instance.
+   *
+   * @param timer the {@link Timer} that is used to trigger the scheduled channel close.
+   *        The recommended {@link Timer} implementation is {@link HashedWheelTimer}.
+   * @param requestIdleTimeSeconds the associated channel will be closed request was performed for the specified
+   *        period of time.  Specify {@code 0} to disable.
+   */
+  public HttpIdleStateHandler(Timer timer, int requestIdleTimeSeconds) {
+    this(timer, requestIdleTimeSeconds, TimeUnit.SECONDS);
+  }
+
+  /**
+   * Creates a new instance.
+   *
+   * @param timer the {@link Timer} that is used to trigger the scheduled event.
+   *        The recommended {@link Timer} implementation is {@link HashedWheelTimer}.
+   * @param requestIdleTime the associated channel will be closed request was performed for the specified
+   *        period of time.  Specify {@code 0} to disable.
+   * @param unit the {@link TimeUnit} of {@code requestIdleTime}, {@code writeIdleTime}, and {@code allIdleTime}
+   */
+  public HttpIdleStateHandler(Timer timer, long requestIdleTime, TimeUnit unit) {
+    if (timer == null) {
+      throw new NullPointerException("timer");
+    }
+    if (unit == null) {
+      throw new NullPointerException("unit");
+    }
+
+    this.timer = timer;
+    if (requestIdleTime <= 0) {
+      requestIdleTimeMillis = 0;
+    } else {
+      requestIdleTimeMillis = Math.max(unit.toMillis(requestIdleTime), 1);
+    }
+  }
+
+  /**
+   * Return the configured requestIdleTime, in milliseconds.
+   *
+   */
+  public long getRequestIdleTimeInMillis() {
+    return requestIdleTimeMillis;
+  }
+
+  /**
+   * Stops the {@link Timer} which was specified in the constructor of this
+   * handler.  You should not call this method if the {@link Timer} is in use
+   * by other objects.
+   */
+  public void releaseExternalResources() {
+    timer.stop();
+  }
+
+  public void beforeAdd(ChannelHandlerContext ctx) throws Exception {
+    if (ctx.getPipeline().isAttached()) {
+      // channelOpen event has been fired already, which means
+      // this.channelOpen() will not be invoked.
+      // We have to initialize here instead.
+      initialize(ctx);
+    } else {
+      // channelOpen event has not been fired yet.
+      // this.channelOpen() will be invoked and initialization will occur there.
+    }
+  }
+
+  public void afterAdd(ChannelHandlerContext ctx) throws Exception {
+    // NOOP
+  }
+
+  public void beforeRemove(ChannelHandlerContext ctx) throws Exception {
+    destroy(ctx);
+  }
+
+  public void afterRemove(ChannelHandlerContext ctx) throws Exception {
+    // NOOP
```

Comment:
Why declare no-op methods? Are they from parent? If that's the case, need to have `@Override`.